Notebook10 Merely averaging

Timeline: 2017/11/15

Goal: Experiment with averaging the results of online kernels

I. Import packages and files

In [71]:
import pandas as pd
import numpy as np

In [72]:
# import filles
print('loading files...')
submission1 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/01submission.csv', na_values=-1)
submission2 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/02submission.csv', na_values=-1)
submission3 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/03submission.csv', na_values=-1)
submission4 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/04submission.csv', na_values=-1)
submission5 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/05submission.csv', na_values=-1)

submission6 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/Online Submissions/01submission.csv', na_values=-1)
submission7 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/Online Submissions/02submission.csv', na_values=-1)
submission8 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/Online Submissions/03submission.csv', na_values=-1)
submission9 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/Online Submissions/04submission.csv', na_values=-1)
submission10 = pd.read_csv('/Users/maxji/Desktop/Kaggle/0SafeDriver/SubmissionsToUse/Online Submissions/05submission.csv', na_values=-1)

loading files...


In [73]:
# print out descriptions to see the distribution
print(submission1.describe())
print(submission2.describe())
print(submission3.describe())
print(submission4.describe())
print(submission5.describe())
print(submission6.describe())
print(submission7.describe())
print(submission8.describe())
print(submission9.describe())
print(submission10.describe())

                 id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.036400
std    4.296830e+05       0.021225
min    0.000000e+00       0.020310
25%    3.720218e+05       0.026990
50%    7.443070e+05       0.031250
75%    1.116308e+06       0.038540
max    1.488026e+06       0.994590
                 id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.036362
std    4.296830e+05       0.020337
min    0.000000e+00       0.020700
25%    3.720218e+05       0.027250
50%    7.443070e+05       0.031400
75%    1.116308e+06       0.038440
max    1.488026e+06       0.986890
                 id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.036362
std    4.296830e+05       0.020337
min    0.000000e+00       0.020700
25%    3.720218e+05       0.027250
50%    7.443070e+05       0.031400
75%    1.116308e+06       0.038440
max    1.488026e+06       0.986890
                 id         target
count  8.928160e+05 

In [45]:
submission4.describe()

id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.056659
std    4.296830e+05       0.029829
min    0.000000e+00       0.011970
25%    3.720218e+05       0.036934
50%    7.443070e+05       0.049850
75%    1.116308e+06       0.067893
max    1.488026e+06       0.629028

In [46]:
submission5.describe()

id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.036411
std    4.296830e+05       0.019759
min    0.000000e+00       0.007053
25%    3.720218e+05       0.023568
50%    7.443070e+05       0.031892
75%    1.116308e+06       0.043726
max    1.488026e+06       0.743323

II. Calculate the average

In [48]:
comb1 = submission4.merge(submission5, how='left', on='id')

In [49]:
comb1.describe()

id       target_x       target_y
count  8.928160e+05  892816.000000  892816.000000
mean   7.441535e+05       0.056659       0.036411
std    4.296830e+05       0.029829       0.019759
min    0.000000e+00       0.011970       0.007053
25%    3.720218e+05       0.036934       0.023568
50%    7.443070e+05       0.049850       0.031892
75%    1.116308e+06       0.067893       0.043726
max    1.488026e+06       0.629028       0.743323

In [68]:
# Having a weighted geometric mean of the selected submissions
final=pd.DataFrame()
final['id'] = submission1['id']
final['target'] = 1
final['target'] *= np.power(submission6['target'],1/4)
final['target'] *= submission7['target']
final['target'] *= submission8['target']
final['target'] *= submission9['target']
final['target'] = np.power(final['target'],4/13)
final.describe()

id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.082690
std    4.296830e+05       0.043002
min    0.000000e+00       0.001599
25%    3.720218e+05       0.053081
50%    7.443070e+05       0.076603
75%    1.116308e+06       0.103943
max    1.488026e+06       0.705899

In [51]:
"""final=submission7.merge(submission8, how='left', on='id')
final=final.merge(submission9,how='left',on='id')
final1 = final.drop(['id'],axis=1)
for i in (0,len(final['target'])-1):
    final['target'][i]=np.median([final['target_x'][i],final['target_y'][i],final['target'][i]])
final.describe()
final = final.drop(['target_x','target_y'],axis=1)
#final['target'] = np.median(submission7['target'],submission8['target'],submission9['target'])
"""

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [52]:
final.describe()

id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.500538
std    4.296830e+05       0.285098
min    0.000000e+00       0.000015
25%    3.720218e+05       0.255402
50%    7.443070e+05       0.500616
75%    1.116308e+06       0.745616
max    1.488026e+06       0.999998

In [56]:
"""final2=submission7.merge(submission8, how='left', on='id')
final2=final2.merge(submission9,how='left',on='id')
print(final2.describe())
final3 = final2.copy().drop(['id'],axis=1)
for i in (0,len(final2['target'])-1):
    final2['target'][i]=np.median(final3['target_x'][i]+final3['target_y'][i]+final3['target'][i])
final2.describe()
final2 = final2.drop(['target_x','target_y'],axis=1)
#final['target'] = np.median(submission7['target'],submission8['target'],submission9['target'])


                 id       target_x       target_y         target
count  8.928160e+05  892816.000000  892816.000000  892816.000000
mean   7.441535e+05       0.036345       0.036559       0.500539
std    4.296830e+05       0.020228       0.020249       0.285097
min    0.000000e+00       0.020739       0.006175       0.000015
25%    3.720218e+05       0.027330       0.023336       0.255403
50%    7.443070e+05       0.031440       0.031776       0.500616
75%    1.116308e+06       0.038419       0.043893       0.745616
max    1.488026e+06       0.988715       0.368518       0.999998


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [58]:
"""final4 = final2.copy()
for i in (0,len(final2['target'])-1):
    final4['target'][i] = final4['target'][i]-final3['target'][i]
print(final4.describe())"""

                 id         target
count  8.928160e+05  892816.000000
mean   7.441535e+05       0.500538
std    4.296830e+05       0.285097
min    0.000000e+00       0.000015
25%    3.720218e+05       0.255402
50%    7.443070e+05       0.500616
75%    1.116308e+06       0.745616
max    1.488026e+06       0.999998


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
"""final=pd.DataFrame()
final['id'] = submission1['id']
final['target'] = 0
final['target'] += (submission1['target'])*4/5
final['target'] += (submission4['target'])/5
final.describe()"""

III. Make submission

In [70]:
final.to_csv('combined1234_weightedaverage14.csv',index=False, float_format='%.5f')

Insight:<br>
This is an example of what people have been doing in online kernels in recent days: directly combining submission files with high score and trying to get a better stacking score. <br>
I tried to change one of the coefficients in the weighing process and get a 0.286 score instead of the original 0.285. But this method is more like random guessing.<br>
Our team decided to stop with this competition once everybody starts desperately doing stacking to improve scores.